File: kmeans.rb

require "matrix"

def kmeans(arr, k)

	c_arr = []
	l_arr = Array.new(arr.length, 0)

	c_arr = arr.sample(k)
	
	50.times do 
		
		c_tmp_arr = Array.new(k){Array.new(arr[0].length, 0)}

		arr.each_with_index do |a, idx_a|
			min_k = 0
			min_dst = -1

			c_arr.each_with_index do |c_a, idx_c_a|
				dst = 0
				c_a.each_with_index do |v, idx_c_v|
					dst += Math.sqrt((v - a[idx_c_v]) ** 2)
				end

				if min_dst > dst || min_dst < 0
					min_dst = dst
					min_k = idx_c_a
				end
			end
			
			#for updating criterion axis
			c_tmp_arr[min_k] = c_tmp_arr[min_k].zip(a).map{|f, s| f + s}
			#update label number
			l_arr[idx_a] = min_k
		end
		
		#update center
		c_tmp_arr.each_with_index do |v, idx|
			new_c = v.map{|a| a / l_arr.count(idx)}	
			c_arr[idx] = new_c
		end	
	end
	puts "ラベルはそれぞれ#{l_arr.to_s}です"
	puts "最終的な中心点は#{c_arr.to_s}です"

	return l_arr, c_arr
end


if __FILE__ == $0
	
	#	sample
	#
	#	argments
	#		arr: N-dim arrray
	#		k: number of split
	#
	
	arr = Matrix.build(20, 3){rand(1..50)}.to_a
	
	l_arr, c_arr = kmeans(arr, 3)
	
	puts l_arr.to_s
	puts c_arr.to_s

end

File: kmeans.js



var kmeans = function(arr, k){
	
	console.log("kmeans start");
	var c_arr = [];
	var l_arr = [];
	var row = arr.length;
	var col = arr[0].length;
	var area = row * col;
	
	//---------------initialize l_arr---------------

	for (var i = 0; i < row; i++){
		var a = [];
		for (var j = 0; j < col; j++){
			a.push(0);
		}
		l_arr.push(a);
	}

	//---------------decide of initial point---------------
	var index_arr = [];
	for (var i = 0; i < row; i++){
		index_arr.push(i);	
	}
	
	for (var i = 0; i < k; i++){
		var r = Math.floor(Math.random() * (index_arr.length - 0) + 0);
		var r_num = index_arr.splice(r, 1);
		var new_c = arr[r_num];
		c_arr.push(new_c);
	}

	//----------------kmeans--------------
	for (var i = 0; i < 50; i++){
		var c_tmp_arr = [];
		for (var j = 0; j < k; j++){
			var a = [];
			for (var l = 0; l < col; l++){
				a.push(0);
			}
			c_tmp_arr.push(a);
		}
		
		for (var j = 0; j < row; j++){
			var a = arr[j];	
			var min_k = 0;
			var min_dst = -1;
			
	//-----------------caluclate distance of center-------------
			for (var l = 0; l < k; l++){
				var dst = 0;
				var c_a = c_arr[l];
				for (var m = 0; m < col; m++){
					var diff = parseInt(c_a[m]) - a[m];
					var c_v = Math.sqrt(Math.pow(diff, 2));
					dst += c_v;
				}
	//------------------update if shorter than min_dst------------
				if (min_dst > dst || min_dst < 0){
					min_dst = dst;
					min_k = l;
				}

			}

	//------------------update c_tmp_arr------------
			var c_tmp_a = c_tmp_arr[min_k];
			for (var n = 0; n < k; n++){
				c_tmp_a[n] += a[n];
			}

			c_tmp_arr[min_k] = c_tmp_a;
			l_arr[j] = min_k;
		}	
		//----------------count elements of each labels--------------
		var num_of_labels = {};

		for (var j = 0; j < row; j++){
			var v = String(l_arr[j])
			if (v in num_of_labels){
				num_of_labels[v] += 1;
			}else{
				num_of_labels[v] = 1;
			}
			
		}
		//update center
		for (var o = 0; o < k; o++){
			var new_c = [];
			for (var p = 0; p < col; p++){
				new_c = c_tmp_arr[o][p] / num_of_labels[o];
				c_arr[o][p] = new_c;
			}
		}
	}
	return [l_arr, c_arr]; 
};


(function(){
	var arr = [];
	for (var i = 0; i < 10; i++){
		var a = [];
		for (var j = 0; j < 3; j++){
			a.push(Math.floor(Math.random() * (50 - 1 + 1)) + 1);
		}
		arr.push(a);
	}

	var dst_arr = kmeans(arr, 2);

	console.log(dst_arr[0]);
	console.log(dst_arr[1]);
})();
